### PR TITLE
chore: replace CoffeeScript examples with JavaScript

### PR DESCRIPTION
**Pull request by chantastic**
This PR brings many things up to date with the way we're building apps today.

* Update all CoffeeScript to use JavaScript with JSTransform-style harmony
* Update all Examples to use JSX
* Update component definition order
* Update examples where recommendation was ambiguous
* Update #JSX section
* Add ES6 Harmony section
* Add note on `addons.classSet` being depricated

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -162,20 +178,22 @@ See: [Cached State in render](#cached-state-in-render) anti-pattern
 
 Name compound state methods with the `is`, `has` or `can` prefix.
 
-```coffeescript
-# bad
-happyAndKnowsIt: ->
-  @state.happy and @state.knowsIt
+```javascript
```
Comment (geolessel) [inline README.md]: The `bad` comment has been lost on this function.

File: README.md [hunk 2]
```diff
@@ -162,20 +178,22 @@ See: [Cached State in render](#cached-state-in-render) anti-pattern
...
+happyAndKnowsIt() {
+  return this.state.happy && this.state.knowsIt;
+},
 
-# good
-isWillingSongParticipant: ->
-  @state.happy and @state.knowsIt
+isWillingSongParticipant() {
```
Comment (geolessel) [inline README.md]: Similar to above, do you want to keep the `good` comment here to be specific and explicit?

File: README.md [hunk 3]
```diff
@@ -188,48 +206,61 @@ See: [Compound Conditions](#compound-conditions) anti-pattern
...
 
-```coffeescript
-# good
-render: ->
-  createItem = (itemText) ->
-    React.DOM.li(null, itemText)
+```javascript
+// bad
+render() {
+  return <div>{this.props.name} {this.state.smiling ? "is smiling" : ""}</div>,
```
Comment (geolessel) [inline README.md]: I know this is the `bad` example, but is the trailing `,` here a typo?

File: README.md [hunk 4]
```diff
@@ -188,48 +206,61 @@ See: [Compound Conditions](#compound-conditions) anti-pattern
...
-render: ->
-  React.DOM.ul(null, @renderItems())
+  return "";
+},
 
-renderItems: ->
-  for itemText in @props.items
-    React.DOM.li(null, itemText)
+render() {
+  return <div>{this.props.name} {this.renderSmilingStatemt()}</div>;
```
Comment (geolessel) [inline README.md]: Your function name is `renderSmilingStatement()`

File: README.md [hunk 5]
```diff
@@ -269,19 +304,39 @@ See: [Computed Props](#computed-props) pattern
...
-  if @isLikeTotallyHappy()
-    React.DOM.div(null, "Knows what it's all about.")
+```javascript
+// bad
+render() {
+  return <div>{if (this.state.happy && this.state.knowsIt) { return "Clapping
+hands" }</div>;
+}
+
+// better
```
Comment (geolessel) [inline README.md]: Is there a good way to visually group these (`bad`, `better`, `betterer`) examples here and in other places? I know one example just goes until the next one starts, but it seems like there could be a good way to further separate them visually.

File: README.md [hunk 6]
```diff
@@ -388,32 +490,40 @@ Child = React.createClass
...
-Person({name: 1337})
-# Warning: Invalid prop `name` of type `number` supplied to `MyValidatedComponent`, expected `string`.
+```html
+<Person name=1337 />
+// Warning: Invalid prop `name` of type `number` supplied to `MyValidatedComponent`, expected `string`.
 ```
 
 Components may require `props`
 
 ```coffeescript
```
Comment (geolessel) [inline README.md]: Change to `javascript` code type.

File: README.md [hunk 7]
```diff
@@ -424,14 +534,19 @@ Read: [Prop Validation](http://facebook.github.io/react/docs/reusable-components
 
 Use Reacts `String.fromCharCode()` for special characters.
```
Comment (geolessel) [inline README.md]: Should be "Use **React's**" (possessive apostrophe).

**Comment by danott** [comment]
LUV IT :+1:

**Comment by chantastic** [comment]
@geolessel you have eyes like a hawk! thanks ;)

I was wondering myself if there was a better way to group things now that the examples themselves are a little more cluttered. maybe split them into two blocks? I'll try that and hit you up for feedback.

**Comment by chantastic** [comment]
@geolessel what are you thoughts on it being in separate blocks? I don't love it but maybe it's better than the alternative https://github.com/planningcenter/react-patterns/tree/javascript-examples#compound-conditions

**Comment by geolessel** [comment]
@chantastic Agreed. Don't love it but it's better than it was.

File: README.md [hunk 8]
```diff
@@ -243,22 +284,28 @@ need to iterate over children and place them within a complex layout.
...
+// good
+render() {
+  return <div>{'Mrs. ' + this.props.name}</div>;
+}
+```
 
-# good (complex example)
-getFormattedBirthDate: ->
-  moment(@props.user.bday).format(LL);
+``javascript
```
Comment (geolessel) [inline README.md]: Missed one backtick.

File: README.md [hunk 9]
```diff
@@ -292,37 +363,73 @@ See: [Compound State](#compound-state) pattern
...
+propTypes: {
+  initialItems: React.PropTypes.array
+},
+
+getInitialState() {
+  return {
+    items: this.props.initialItems
+  };
+
+// best (complex)
```
Comment (geolessel) [inline README.md]: Do you want to split `best` from `good` here as well? I'm fine either way but for consistency I think perhaps we should split it.

File: README.md (hunk 8 above)
Comment (chantastic) [inline README.md]: weird. i `yanked/put` all those ;)

**Comment by geolessel** [comment]
:+1: After final changes (mainly the missing backtick). Nice job @chantastic!

File: README.md (hunk 9 above)
Comment (chantastic) [inline README.md]: there be some craziness there. i don't think that example should even be there. vim'ing too hard.

**Comment by chantastic** [comment]
Thanks!

**Pull request merged**